discounts/rust/discounts/default/src/run.rs

```diff
@@ -0,0 +1,391 @@
+
+#[derive(Deserialize)]
+struct Metafield {
+    cart_percent: Option<Decimal>,
+    product_percent: Option<Decimal>,
+    delivery_percent: Option<Decimal>,
+}
+
+#[shopify_function_target(
+    target = "cart_run",
+    query_path = "src/run.graphql",
+    schema_path = "schema.graphql"
+)]
+fn cart_run(input: CartResponseData) -> Result<FunctionCartRunResult> {
+    let default = FunctionCartRunResult { operations: vec![] };
+
+    let fetch_result = input.fetch_result.as_ref().expect("Missing fetch result");
+    if fetch_result.status != 200 {
+        return Ok(default);
+    }
+
+    let metafield = input
+        .discount_node
+        .metafield
+        .expect("Missing required metafield configuration");
+    let metafield_value = string_to_config::<Metafield>(&metafield.value);
+
+    let mut operations: Vec<CartOperation> = vec![];
+
+    if metafield_value.cart_percent.is_some() {
+        operations.push(CartOperation::AddOrderDiscounts(OrderDiscounts {
+            selection_strategy: OrderDiscountSelectionStrategy::FIRST,
+            candidates: vec![OrderDiscountCandidate {
+                targets: vec![OrderDiscountCandidateTarget::OrderSubtotal(
+                    OrderSubtotalTarget {
+                        excluded_variant_ids: vec![],
+                    },
+                )],
+                associated_discount_code: None,
+                message: None,
+                value: OrderDiscountCandidateValue::Percentage(CartPercentage {
+                    value: metafield_value.cart_percent.unwrap(),
+                }),
+                conditions: None,
+            }],
+        }));
+    }
+
+    if metafield_value.product_percent.is_some() {
+        let highest_priced_line = input.cart.lines.iter().max_by(|a, b| {
+            let a_amount = a.cost.subtotal_amount.amount;
+            let b_amount = b.cost.subtotal_amount.amount;
+            a_amount.partial_cmp(&b_amount).unwrap()
+        });
+        operations.push(CartOperation::AddProductDiscounts(ProductDiscounts {
+            selection_strategy: ProductDiscountSelectionStrategy::FIRST,
+            candidates: vec![ProductDiscountCandidate {
+                targets: vec![ProductDiscountCandidateTarget::CartLine(CartLineTarget {
+                    id: highest_priced_line.unwrap().id.clone(),
+                    quantity: Some(1),
+                })],
+                associated_discount_code: None,
+                message: None,
+                value: ProductDiscountCandidateValue::Percentage(CartPercentage {
+                    value: metafield_value.product_percent.unwrap(),
+                }),
+            }],
+        }));
+    }
+
+    Ok(FunctionCartRunResult { operations })
+}
+
+#[shopify_function_target(
+    target = "delivery_run",
+    query_path = "src/run.graphql",
+    schema_path = "schema.graphql"
+)]
+fn delivery_run(input: DeliveryResponseData) -> Result<FunctionDeliveryRunResult> {
```

Comment:
This `delivery_run` will make:
- no discount if fetch did not return a 200. As above, I could check if 401 or the body.
- a Z% discount on all delivery groups